## aerobic-anaerobic-genomics.R by Rohan Maddamsetti.
## examine evolution in aerobic- and anaerobic-specific genes
## in the 50000 generation LTEE genomes.

## for details on arcA mutation timing, see: http://barricklab.org/shiny/LTEE-Ecoli/
## and Ben Good's dataset.

library(tidyverse)

home.dir <- path.expand("~")
## If on Nkrumah's computer, set working dir to ~/Desktop/ArcA_analysis.
if (str_detect(home.dir,'nkrumahgrant')) {
    setwd(file.path(home.dir,"Desktop/ArcA_analysis/"))
}

parse.Salmon.etal.data <- function(datafile) {
    
    ## Orthologs were pulled from the OMA database between
    ## REL606 and
    ## Escherichia coli (strain K12 / MC4100 / BW2952) (ECOBW).
    ## MC4100 is the arcA+ strain for which I have gene expression data.
    dataf <- read.csv(datafile, as.is=TRUE)
    
    OMA.gps <- read.csv("../data/OMApairs_REL606_MC4100.csv", as.is=TRUE) %>%
        filter(orthology_mapping == "1:1")
    
    ## MC4100_IDs.csv was generated by running:
    ## python printEcoliIDs.py -i CP001396.gbk > MC4100_IDs.csv
    MC4100.ids <- read.csv('../results/MC4100_IDs.csv', header = TRUE, as.is=TRUE) %>%
        mutate(MC4100_locus_tag = locus_tag) %>% 
        left_join(OMA.gps)
    
    ## Ecocyc_gene_to_blattner.csv was generated by make_gene_to_blattner_csv.py.
    ecocyc.gene.to.blattner <- read.csv('../data/Ecocyc_gene_to_blattner.csv',
                                        header = TRUE,
                                        stringsAsFactors = FALSE) %>%
        mutate(Locus = gene) %>% select(-gene)
    
    my.data <- dataf %>%
        mutate(Locus = Gene.Name) %>% select(-Gene.Name) %>%
        left_join(ecocyc.gene.to.blattner) %>%
        filter(!is.na(blattner)) %>%
        left_join(MC4100.ids) %>%
        drop_na() %>%
        filter(PPDE.p. > 0.99) %>%
        select(Bayesian_p, Bayesian_lnp, PPDE..p.,
               PPDE.p., fold, Locus, blattner, Gene, locus_tag, REL606_locus_tag, gene_length)
    
    return(my.data)
}

O2.data <- parse.Salmon.etal.data("../data/O2ExpressionK12.csv")
## fold-change is the ratios of the means of the measurements of K-12 str. MC4100
## grown under aerobic or anaerobic conditions.
## A negative sign implies decreased expression in strain MC4100 grown under
## anaerobic conditions.
## genes with negative fold changes should be aerobic genes.

up.O2.data <- filter(O2.data,fold>0)
down.O2.data <- filter(O2.data,fold<0)

mutator.50K <- read.csv('../data/Gen50000_M.csv', header=TRUE, as.is=TRUE) %>%
filter(clone=='A')

nonmutator.50K <- read.csv('../data/Gen50000_NM.csv', header=TRUE, as.is=TRUE) %>%
filter(clone=='A')

mutator.50K.in.up.O2.data <- mutator.50K %>%
filter(locus_tag %in% up.O2.data$REL606_locus_tag)

mutator.50K.in.down.O2.data <- mutator.50K %>%
filter(locus_tag %in% down.O2.data$REL606_locus_tag)

nonmutator.50K.in.up.O2.data <- nonmutator.50K %>%
filter(locus_tag %in% up.O2.data$REL606_locus_tag)

nonmutator.50K.in.down.O2.data <- nonmutator.50K %>%
filter(locus_tag %in% down.O2.data$REL606_locus_tag)

## fold-change is the ratios of the means of the measurements of K-12 str. MC4100
## grown under aerobic or anaerobic conditions.
## A negative sign implies decreased expression in strain MC4100 grown under
## anaerobic conditions.
## genes with negative fold changes should be aerobic genes.

## target.prob = 0.676. So anaerobic genes are a bigger target.
target.prob <- sum(up.O2.data$gene_length)/(sum(up.O2.data$gene_length)+sum(down.O2.data$gene_length))

## aerobic genes are under positive selection.
nrow(nonmutator.50K.in.up.O2.data) ## 21
nrow(nonmutator.50K.in.down.O2.data) ## 38
## highly significant difference: p-value = 8.39e-07.
binom.test(x=nrow(nonmutator.50K.in.up.O2.data),
           n = nrow(nonmutator.50K.in.up.O2.data)+nrow(nonmutator.50K.in.down.O2.data),
            p=target.prob, alternative="two.sided")

## Nkrumah predicts that anaerobic-specific genes accumulate more mutations than
## aerobic specific genes, due to relaxed selection.

nrow(mutator.50K.in.up.O2.data) ## 836
nrow(mutator.50K.in.down.O2.data) ## 333

## still significant in mutators!! p-value = 0.004036.
binom.test(x = nrow(mutator.50K.in.up.O2.data),
           n = nrow(mutator.50K.in.up.O2.data)+nrow(mutator.50K.in.down.O2.data),
           p = target.prob, alternative="two.sided")


mutator.50K.in.up.O2.data.no.dS <- mutator.50K.in.up.O2.data %>%
    filter(snp_type != 'synonymous')

mutator.50K.in.down.O2.data.no.dS <- mutator.50K.in.down.O2.data %>%
    filter(snp_type != 'synonymous')

nrow(mutator.50K.in.up.O2.data.no.dS) ## 644
nrow(mutator.50K.in.down.O2.data.no.dS) ## 251

## still significant in mutators!! p-value = 0.005317.
binom.test(x = nrow(mutator.50K.in.up.O2.data.no.dS),
           n = nrow(mutator.50K.in.up.O2.data.no.dS)+nrow(mutator.50K.in.down.O2.data.no.dS),
           p = target.prob, alternative="two.sided")

## redo, just on non-synonymous mutations.
## non-mutator result is more significant! p < 10^-8.
binom.test(x = nrow(filter(nonmutator.50K.in.up.O2.data,snp_type=='nonsynonymous')),
           n = nrow(filter(nonmutator.50K.in.up.O2.data,snp_type=='nonsynonymous'))+nrow(filter(nonmutator.50K.in.down.O2.data,snp_type=='nonsynonymous')),
           p=target.prob,
           alternative="two.sided")

## mutator result, using gene length as target size,
## just considering dN-- result is insignificant (p = 0.1113).
mutator.50K.in.up.O2.data.just.dN <- filter(mutator.50K.in.up.O2.data, snp_type == 'nonsynonymous')
mutator.50K.in.down.O2.data.just.dN <- filter(mutator.50K.in.down.O2.data, snp_type == 'nonsynonymous')

binom.test(x = nrow(mutator.50K.in.up.O2.data.just.dN),
           n = nrow(mutator.50K.in.up.O2.data.just.dN)+nrow(mutator.50K.in.down.O2.data.just.dN),
           p = target.prob,
           alternative="two.sided")

## mutator result, using gene length as target size,
## more significant when considering dS, but still not significant (p = 0.0672).
mutator.50K.in.up.O2.data.dNanddS <- filter(mutator.50K.in.up.O2.data, snp_type %in% c('nonsynonymous','synonymous'))
mutator.50K.in.down.O2.data.dNanddS <- filter(mutator.50K.in.down.O2.data, snp_type %in% c('nonsynonymous','synonymous'))

binom.test(x = nrow(mutator.50K.in.up.O2.data.dNanddS),
           n = nrow(mutator.50K.in.up.O2.data.dNanddS)+nrow(mutator.50K.in.down.O2.data.dNanddS),
           p = target.prob,
           alternative="two.sided")

## And more significant again when including (rare) nonsense mutations,
## but still marginally insignficant (p = 0.0525).
mutator.50K.in.up.O2.data.just.pointmut <- mutator.50K.in.up.O2.data %>%
    filter(snp_type != 'not_snp')

mutator.50K.in.down.O2.data.just.pointmut <- mutator.50K.in.down.O2.data %>%
    filter(snp_type != 'not_snp')

binom.test(x = nrow(mutator.50K.in.up.O2.data.just.pointmut),
           n = nrow(mutator.50K.in.up.O2.data.just.pointmut)+nrow(mutator.50K.in.down.O2.data.just.pointmut),
           p = target.prob,
           alternative="two.sided")

##################
## write out data files for metagenomics analysis.
## we want anaerobic-specific and aerobic-specific genes.
write.csv(select(up.O2.data,blattner,Gene,REL606_locus_tag),file="../results/anaerobic-specific-genes.csv")
write.csv(select(down.O2.data,blattner,Gene,REL606_locus_tag),file="../results/aerobic-specific-genes.csv")
