#This is the analysis script for determining if the number of mutations in anaerobic
#related genes have less mutations than expected by chance.

## for details on arcA mutation timing, see: http://barricklab.org/shiny/LTEE-Ecoli/
## and Ben Good's dataset.

library(tidyverse)

home.dir <- path.expand("~")
## If on Nkrumah's computer, set working dir to ~/Desktop/ArcA_analysis.
if (str_detect(home.dir,'nkrumahgrant')) {
    setwd(file.path(home.dir,"Desktop/ArcA_analysis/"))
}

#################
## Is there overdispersion/anti-correlation between mutations in
## arcA and arcB? Not when looking at just 50K genomes.
## p = 0.2424 by Fisher's exact test.
## 7/12 50K clones have arcA mutations.
## 6/12 50K clones have arcB mutations.
## 2/12 50K clones have both arcA and arcB mutations.

## Fisher's exact test on 50K clones:
## 2 have both arcA and arcB.
## 5 have arcA but not arcB.
## 4 have arcB but not arcA.
## 1 has neither.

fisher.test(matrix(c(2,4,5,1),nrow=2))

#################
## tests on aerobic- and anaerobic-specific genes.

parse.Salmon.etal.data <- function(dataf) {

  ## Orthologs were pulled from the OMA database between
  ## REL606 and
  ## Escherichia coli (strain K12 / MC4100 / BW2952) (ECOBW).
  ## MC4100 is the arcA+ strain for which I have gene expression data.
  OMA.gps <- read.csv("../data/OMApairs_REL606_MC4100.csv", stringsAsFactors = FALSE) %>%
  filter(orthology_mapping == "1:1")

  ## MC4100_IDs.csv was generated by running:
  ## python printEcoliIDs.py -i CP001396.gbk > MC4100_IDs.csv
  MC4100.ids <- read.csv('../data/MC4100_IDs.csv', stringsAsFactors = FALSE, header = TRUE) %>%
  mutate(MC4100_locus_tag = locus_tag) %>% 
  left_join(OMA.gps)
  
  ## Ecocyc_gene_to_blattner.csv was generated by make_gene_to_blattner_csv.py.
  ecocyc.gene.to.blattner <- read.csv('../data/Ecocyc_gene_to_blattner.csv',
                                      header = TRUE,
                                      stringsAsFactors = FALSE) %>%
                                      mutate(Locus = gene) %>% select(-gene)
  
  my.data <- read.csv(dataf, header = TRUE, stringsAsFactors = FALSE) %>%
  mutate(Locus = Gene.Name) %>% select(-Gene.Name) %>%
  left_join(ecocyc.gene.to.blattner) %>%
  filter(!is.na(blattner)) %>%
  left_join(MC4100.ids) %>%
  drop_na() %>%
  filter(PPDE.p. > 0.99) %>%
  select(Bayesian_p, Bayesian_lnp, PPDE..p.,
         PPDE.p., fold, Locus, blattner, gene, locus_tag, REL606_locus_tag, length)

  return(my.data)
}

O2.data <- parse.Salmon.etal.data("../data/O2ExpressionK12.csv")
## fold-change is the ratios of the means of the measurements of K-12 str. MC4100
## grown under aerobic or anaerobic conditions.
## A negative sign implies decreased expression in strain MC4100 grown under
## anaerobic conditions.
## genes with negative fold changes should be aerobic genes.

up.O2.data <- filter(O2.data,fold>0)
down.O2.data <- filter(O2.data,fold<0)

## to look at Ben Good's dataset we want to get the list of anaerobic-specific and
## aerobic-specific genes.
write.csv(select(up.O2.data,blattner,gene,REL606_locus_tag),file="../results/anaerobic-specific-genes.csv")
write.csv(select(down.O2.data,blattner,gene,REL606_locus_tag),file="../results/aerobic-specific-genes.csv")

mutator.50K <- read.csv('../data/Gen50000_M.csv',header=TRUE,stringsAsFactors=FALSE) %>%
  select(population,strain,clone,mutator_status,type,start_position,end_position,gene_position,
         locus_tag,mutation_category,snp_type) %>% filter(clone=='A')

nonmutator.50K <- read.csv('../data/Gen50000_NM.csv',header=TRUE,stringsAsFactors=FALSE) %>%
  select(population,strain,clone,mutator_status,type,start_position,end_position,gene_position,
         locus_tag,mutation_category,snp_type) %>% filter(clone=='A')


mutator.50K.in.up.O2.data <- mutator.50K %>% filter(locus_tag %in% up.O2.data$REL606_locus_tag)
mutator.50K.in.down.O2.data <- mutator.50K %>% filter(locus_tag %in% down.O2.data$REL606_locus_tag)

## originally I filtered out synonymous mutations to reduce noise, but not necessary!
mutator.50K.in.up.O2.data.no.dS <- filter(mutator.50K.in.up.O2.data, snp_type != 'synonymous')
mutator.50K.in.down.O2.data.no.dS <- filter(mutator.50K.in.down.O2.data, snp_type != 'synonymous')

nonmutator.50K.in.up.O2.data <- nonmutator.50K %>% filter(locus_tag %in% up.O2.data$REL606_locus_tag)
nonmutator.50K.in.down.O2.data <- nonmutator.50K %>% filter(locus_tag %in% down.O2.data$REL606_locus_tag)

## fold-change is the ratios of the means of the measurements of K-12 str. MC4100
## grown under aerobic or anaerobic conditions.
## A negative sign implies decreased expression in strain MC4100 grown under
## anaerobic conditions.
## genes with negative fold changes should be aerobic genes.

## target.prob = 0.676. So anaerobic genes are a bigger target.
target.prob <- sum(up.O2.data$length)/(sum(up.O2.data$length)+sum(down.O2.data$length))

## aerobic genes are under positive selection.
nrow(nonmutator.50K.in.up.O2.data) ## 21
nrow(nonmutator.50K.in.down.O2.data) ## 38
## highly significant difference: p-value = 8.39e-07.
binom.test(x=nrow(nonmutator.50K.in.up.O2.data), n = nrow(nonmutator.50K.in.up.O2.data)+nrow(nonmutator.50K.in.down.O2.data),
            p=target.prob,alternative="two.sided")

## Nkrumah predicts that anaerobic-specific genes accumulate more mutations than
## aerobic specific genes, due to relaxed selection.

nrow(mutator.50K.in.up.O2.data) ## 836
nrow(mutator.50K.in.down.O2.data) ## 333

## still significant in mutators!! p-value = 0.004036.
binom.test(x=nrow(mutator.50K.in.up.O2.data), n = nrow(mutator.50K.in.up.O2.data)+nrow(mutator.50K.in.down.O2.data),
            p=target.prob,alternative="two.sided")

nrow(mutator.50K.in.up.O2.data.no.dS) ## 644
nrow(mutator.50K.in.down.O2.data.no.dS) ## 244

## still significant in mutators!! p-value = 0.005317.
binom.test(x=nrow(mutator.50K.in.up.O2.data.no.dS), n = nrow(mutator.50K.in.up.O2.data.no.dS)+nrow(mutator.50K.in.down.O2.data.no.dS),
            p=target.prob,alternative="two.sided")

## redo, just on non-synonymous mutations, and on rigorously computed target sizes
## from measureTargetSize.py.

## numbers gotten by running measureTargetSize.py.
## Use these to normalize cumulative mutations over time.
anaerobic.synon.sites <- 103549
anaerobic.nonsynon.sites <- 343951
aerobic.synon.sites <- 50022.66666666666
aerobic.nonsynon.sites <- 168705.33333333337
total.synon.sites <- 890787
total.nonsynon.sites <- 3001647

target.prob2 <- anaerobic.nonsynon.sites/(anaerobic.nonsynon.sites+aerobic.nonsynon.sites)
target.prob3 <- (anaerobic.nonsynon.sites+anaerobic.synon.sites)/(anaerobic.nonsynon.sites+anaerobic.synon.sites+aerobic.nonsynon.sites+aerobic.synon.sites)

## non-mutator result is more significant! p < 10^-8.
binom.test(x=nrow(filter(nonmutator.50K.in.up.O2.data,snp_type=='nonsynonymous')), n = nrow(filter(nonmutator.50K.in.up.O2.data,snp_type=='nonsynonymous'))+nrow(filter(nonmutator.50K.in.down.O2.data,snp_type=='nonsynonymous')),
            p=target.prob,alternative="two.sided")

## By using a more accurate target distribution, it's clear that this effect is driven by
## non-point mutations. There's no significant reduction in hits if we JUST consider non-synonymous mutations.

## just considering dN-- result is marginally insignificant (p = 0.05893).
mutator.50K.in.up.O2.data.just.dN <- filter(mutator.50K.in.up.O2.data, snp_type == 'nonsynonymous')
mutator.50K.in.down.O2.data.just.dN <- filter(mutator.50K.in.down.O2.data, snp_type == 'nonsynonymous')

binom.test(x=nrow(mutator.50K.in.up.O2.data.just.dN), n = nrow(mutator.50K.in.up.O2.data.just.dN)+nrow(mutator.50K.in.down.O2.data.just.dN),
            p=target.prob2,alternative="two.sided")

## my guess is that dS will reduce this signal-- but I'm wrong! (p = 0.03417)
mutator.50K.in.up.O2.data.dNanddS <- filter(mutator.50K.in.up.O2.data, snp_type %in% c('nonsynonymous','synonymous'))
mutator.50K.in.down.O2.data.dNanddS <- filter(mutator.50K.in.down.O2.data, snp_type %in% c('nonsynonymous','synonymous'))

binom.test(x=nrow(mutator.50K.in.up.O2.data.dNanddS), n = nrow(mutator.50K.in.up.O2.data.dNanddS)+nrow(mutator.50K.in.down.O2.data.dNanddS),
            p=target.prob3,alternative="two.sided")


##################
## Look at upregulated in ArcA knockout (KO) and downregulated in ArcA KO.
## These data were collected in anaerobic conditions.
arcA.KO.data <- parse.Salmon.etal.data("../data/ArcAExpressionK12.csv")

## fold-change is the ratios of the means of the measurements of K-12 str. MC4100
## with arcA- divided by arcA+ in anaerobic conditions.
## A negative sign implies decreased expression in arcA-.
## A positive sign implies increased expression in arcA-.

up.arcA.KO.data <- filter(arcA.KO.data,fold>0)
down.arcA.KO.data <- filter(arcA.KO.data,fold<0)

## to look at Ben Good's dataset we want to get the list of
## upregulated and downregulated genes after arcA KO in anaerobic conditions.

write.csv(select(up.arcA.KO.data,blattner,gene,REL606_locus_tag),
          file="../results/arcA-KO-upregulated-genes.csv")
write.csv(select(down.arcA.KO.data,blattner,gene,REL606_locus_tag),
          file="../results/arcA-KO-downregulated-genes.csv")
